src/main/java/io/gravitee/policy/apikey/ApiKeyPolicy.java

```diff
@@ -15,138 +15,172 @@
-import io.gravitee.gateway.api.ExecutionContext;
-import io.gravitee.gateway.api.Request;
-import io.gravitee.gateway.api.Response;
-import io.gravitee.policy.api.PolicyChain;
-import io.gravitee.policy.api.PolicyResult;
-import io.gravitee.policy.api.annotations.OnRequest;
+import io.gravitee.gateway.reactive.api.ExecutionFailure;
+import io.gravitee.gateway.reactive.api.context.ExecutionContext;
+import io.gravitee.gateway.reactive.api.context.Request;
+import io.gravitee.gateway.reactive.api.context.RequestExecutionContext;
+import io.gravitee.gateway.reactive.api.policy.SecurityPolicy;
 import io.gravitee.policy.apikey.configuration.ApiKeyPolicyConfiguration;
-import io.gravitee.repository.exceptions.TechnicalException;
+import io.gravitee.policy.v3.apikey.ApiKeyPolicyV3;
 import io.gravitee.repository.management.api.ApiKeyRepository;
 import io.gravitee.repository.management.model.ApiKey;
+import io.reactivex.Completable;
+import io.reactivex.Single;
 import java.util.Date;
 import java.util.Optional;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.springframework.core.env.Environment;
 
 /**
  * @author David BRASSELY (david.brassely at graviteesource.com)
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
  * @author GraviteeSource Team
  */
-@SuppressWarnings("unused")
-public class ApiKeyPolicy {
+public class ApiKeyPolicy extends ApiKeyPolicyV3 implements SecurityPolicy {
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(ApiKeyPolicy.class);
+    private static final Logger log = LoggerFactory.getLogger(ApiKeyPolicy.class);
 
     static final String ATTR_API_KEY = ExecutionContext.ATTR_PREFIX + "api-key";
 
-    private static final String API_KEY_MISSING_KEY = "API_KEY_MISSING";
-    private static final String API_KEY_INVALID_KEY = "API_KEY_INVALID";
-
-    /**
-     * Policy configuration
-     */
-    private final ApiKeyPolicyConfiguration apiKeyPolicyConfiguration;
+    protected static final String API_KEY_MISSING_KEY = "API_KEY_MISSING";
+    protected static final String API_KEY_INVALID_KEY = "API_KEY_INVALID";
 
     static String API_KEY_HEADER, API_KEY_QUERY_PARAMETER;
     static final String API_KEY_HEADER_PROPERTY = "policy.api-key.header";
     static final String API_KEY_QUERY_PARAMETER_PROPERTY = "policy.api-key.param";
     static final String DEFAULT_API_KEY_QUERY_PARAMETER = "api-key";
     static final String DEFAULT_API_KEY_HEADER_PARAMETER = GraviteeHttpHeader.X_GRAVITEE_API_KEY;
 
-    public ApiKeyPolicy(ApiKeyPolicyConfiguration apiKeyPolicyConfiguration) {
-        this.apiKeyPolicyConfiguration = apiKeyPolicyConfiguration;
+    private final boolean propagateApiKey;
+
+    public ApiKeyPolicy(ApiKeyPolicyConfiguration configuration) {
+        super(configuration);
+        this.propagateApiKey = configuration != null && configuration.isPropagateApiKey();
+    }
+
+    @Override
+    public String id() {
+        return "api-key";
```

Comment:
```suggestion
        return "policy-api-key";
```

Reply:
Actually, the id is aligned with the one defined in the manifest (see https://github.com/gravitee-io/gravitee-policy-apikey/blob/prototype-reactive-security-chain/src/main/resources/plugin.properties).
For now, it's preferable to keep it sync with manifest.